lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,351 @@
+            endpointStatusFlow.collect { endpointMap ->
+                checkAndInitiateConnections(endpointMap)
+            }
+        }
+    }
+
+    private fun checkAndInitiateConnections(endpointMap: Map<String, EndpointInfo>) {
+        val connectedOutgoing = endpointMap.values.count { it.status == EndpointStatus.CONNECTED && it.isOutgoing }
+        val disconnectedEndpoints = endpointMap.values.filter { it.status == EndpointStatus.DISCONNECTED }
+
+        log( "Connected outgoing: $connectedOutgoing, Disconnected endpoints: ${disconnectedEndpoints.size}")
+
+        if (connectedOutgoing < desiredOutgoingConnections && disconnectedEndpoints.isNotEmpty()) {
+            val endpointsToConnect = disconnectedEndpoints.take(desiredOutgoingConnections - connectedOutgoing)
+            log( "Initiating connection to ${endpointsToConnect.size} endpoints")
+            updateMultipleEndpointsStatus(endpointsToConnect.map { it.endpointId }, EndpointStatus.CONNECTING)
+            endpointsToConnect.forEach { requestConnection(it.endpointId) }
+        }
+    }
+
+    private fun updateMultipleEndpointsStatus(endpointIds: List<String>, status: EndpointStatus) {
+        _endpointStatusFlow.update { currentMap ->
+            currentMap.toMutableMap().apply {
+                endpointIds.forEach { endpointId ->
+                    this[endpointId]?.status = status
+                    this[endpointId]?.isOutgoing = true
+                }
+            }
+        }
+    }
+
+    private fun sendMmcpPingPacket(endpointId: String) {
+        val mmcpPing = MmcpPing(Random.nextInt())
+        val pingBytes = mmcpPing.toBytes()
+
+        val header = VirtualPacketHeader(
+            fromAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            toAddr = InetAddress.getByName(broadcastAddress).address.ip4AddressToInt(),
+            fromPort = 0,
+            toPort = 0,
+            lastHopAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            payloadSize = pingBytes.size,
+            hopCount = 0,
+            maxHops = 1
+        )
+
+        val totalSize = VirtualPacketHeader.HEADER_SIZE + pingBytes.size
+        val data = ByteArray(totalSize)
+        header.toBytes(data, 0)
+        System.arraycopy(pingBytes, 0, data, VirtualPacketHeader.HEADER_SIZE, pingBytes.size)
+
+        val virtualPacket = VirtualPacket.fromData(data, 0)
+        val payload = Payload.fromBytes(virtualPacket.data)
+
+        connectionsClient.sendPayload(endpointId, payload)
+            .addOnSuccessListener { log("Sent MMCP Ping to $endpointId") }
+            .addOnFailureListener { e -> log("Failed to send MMCP Ping to $endpointId", e) }
+    }
+
+    private fun sendMmcpPongPacket(endpointId: String, replyToMessageId: Int) {
+        val mmcpPong = MmcpPong(Random.nextInt(), replyToMessageId)
+        val pongBytes = mmcpPong.toBytes()
+
+        val header = VirtualPacketHeader(
```

Comment:
As above - change to using MmcpMessage's .toVirtualPacket, don't use system arraycopy